src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCLKernels.scala

```diff
@@ -0,0 +1,252 @@
+
+      val softmaxGradKernel = cudaSoftmaxGrad[Float](true) /* Default is taking log */
+
+      FunOps7(softmaxGradKernel).apply(
+        Wrap[Array[Float]](dinput.x), 
+        Wrap[Array[Float]](doutput_tensor),
+        Wrap[Array[Float]](output_tensor),
+        unit[Int](last_dim_size),
+        dim3(unit[Int](outer_size)),
+        dim3(unit[Int](1024)),
+        unit[Int](1024 * 4)
+      )
+
+      dinput.x
+    
+    case Node(s, "tensor_transpose", Backend.Const(tt:TensorType)::Backend.Const(anno:Anno)::(operand:Backend.Sym)::_, _) =>
+      val sourceTensor = new TENSOR(s, useOldMetadata = true)
+
+      implicit val sc_ : SourceContext = sourceTensor.pos
+      val m = sourceTensor.et
+
+      val input_shape = tensor_shape(operand, useOldMetadata = true)
+      val input_tensor = get_operand(operand, anno)
+      /*
+      val n_rows = input_shape(0) / 2 // Luke: HARDCODE!
+      val n_cols = input_shape(1)
+      
+      val size = input_shape.fold(1) { _ * _ }
+      val output = gpu_array(size, manifest[Float], myNCCLRank)
+
+      val tileDim = 32
+      val blockRows = 8
+      
+      val transposeKernel = cudaTranspose[Float]
+      FunOps6(transposeKernel).apply(
+        Wrap[Array[Float]](input_tensor),
+        Wrap[Array[Float]](output.x),
+        unit[Int](n_rows),
+        unit[Int](n_cols),
+        dim3(unit[Int]((n_cols + tileDim - 1) / tileDim), unit[Int]((n_rows + tileDim - 1) / tileDim)),
+        dim3(unit[Int](tileDim), unit[Int](blockRows)))
+      
+      output.x
+      */
+      anno match {
+        case NAnno => throw new Exception(s"TODO: not yet handling NAnno")
+        case SAnno(dim: Dim, devices: Seq[Device], _) if tt.contains(dim) =>
+          val shape = tt.shapeSizeAfterSplit(dim, devices.size)
+          val count = numeral(shape)
+          val n_rows = shape(1)
+          val n_cols = shape(0)
+          val output = gpu_array(count, manifest[Float], myNCCLRank)
+          val tileDim = 32
+          val blockRows = 8
+          val transposeKernel = cudaTranspose[Float]
+          FunOps6(transposeKernel).apply(
+            Wrap[Array[Float]](input_tensor),
+            Wrap[Array[Float]](output.x),
+            unit[Int](n_rows),
+            unit[Int](n_cols),
+            dim3(unit[Int]((n_cols + tileDim - 1) / tileDim), unit[Int]((n_rows + tileDim - 1) / tileDim)),
+            dim3(unit[Int](tileDim), unit[Int](blockRows)))
+      
+          output.x
+
+        case SAnno(dim: Dim, devices: Seq[Device], _) => throw new Exception(s"TODO: not yet handling SAnno with AllReduce")
+        case a => throw new Exception(s"TODO: annotation $a is not yet handled")
+      }
+
+
+    case Node(s, "tensors_split2", Backend.Const(tts: List[TensorType])::Backend.Const(anno:Anno)::(input:Backend.Sym)::Backend.Const(axis:Int)::_, _) =>
+      val oldSplitOp = new TENSORS(s, useOldMetadata = true)
+      implicit val sc_ : SourceContext = oldSplitOp.p
```

Comment:
I see you spent effort adding "tensor_split2" but we are not using this op (instead, we use "tensor_split", which can split in any number of results). Do you want to remove this from the PR or just leave a comment to say that this op is deprecated?

Reply:
actually split/concat is still WIP, the split2 is just a temporary name, can we merge after I finish implementing split/concat?

Reply:
sure